### ISSUE TITLE
One Cycle policy in Ludwig

### ISSUE THREAD

**Issue by lucaventurini** [ludwig-ai/ludwig#589]
Hi @w4nderlust ,

is there any plan to implement Smith's One Cycle policy in the learning rate scheduler? (https://arxiv.org/pdf/1803.09820.pdf)

I see that currently the default policy is to anneal the LR (or the batch size). The One Cycle scheduler has shown very good results since it was introduced, and it seems much faster to converge than other schedulers.

**Comment by w4nderlust** [comment]
Why don't you attempt a PR about it? :) All the relevant code is in the train function: https://github.com/uber/ludwig/blob/master/ludwig/models/model.py#L267

**Comment by lucaventurini** [comment]
I don't have much familiarity with tensorflow unfortunately 😅 

Also, in pytorch it's already implemented (https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR), but in TF I wouldn't know where to start, I can't find the equivalent.

**Comment by lordeddard** [comment]
@w4nderlust I am willing to work on this, if no one isn't.

**Comment by w4nderlust** [comment]
@lordeddard that would be great. For starters i suggest you to look at the train function of model.py https://github.com/uber/ludwig/blob/master/ludwig/models/model.py#L293 and look at how the learning rate is treated there. Once that is clear we can discuss how to implement the one cycle policy.

**Comment by lordeddard** [comment]
Hi @w4nderlust , was a bit busy last week. Yes, I will take a look at that and also read a bit more about one cycle policy and discuss the implementation. Thanks!

**Comment by lordeddard** [comment]
Hello @w4nderlust , I have taken a look at the train function in `model.py`. The learning rate for every batch/step is calculated in this [line](https://github.com/uber/ludwig/blob/4039f93e605ed929e4f4a75c725bb70e148df92b/ludwig/models/model.py#L544). An the actual value of the learning rate will be from [here](https://github.com/uber/ludwig/blob/71f958205604e8ab6d4133df9b53ccb50b7073cc/ludwig/utils/math_utils.py#L85). I am presuming we have add code in the same file to calculate the learning rate.

**Comment by w4nderlust** [comment]
@lordeddard yes it may look similar, assuming there is a one_cycle_policy parameter in the training configuration I imagine to have in the model.train() function something that looks like
```
if one_cycle_policy:
  current_learning_rate = one_cycle_policy(previous_learning_rate, curr_epoch, ... other parameters)
```
